File: src/main/java/destiny/iching/Symbol.java
/**
 * @author smallufo
 * Created on 2002/8/13 at 下午 01:08:17
 */
package destiny.iching;

import destiny.core.chinese.FiveElement;
import destiny.core.chinese.FiveElementIF;
import destiny.core.chinese.YinYangIF;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;


/**
 * 八卦基本符號以及其資料
 */
public enum Symbol implements Serializable , SymbolIF , FiveElementIF
{
  乾('乾', new boolean[] {true  , true  , true  } ),
  兌('兌', new boolean[] {true  , true  , false } ),
  離('離', new boolean[] {true  , false , true  } ),
  震('震', new boolean[] {true  , false , false } ),
  巽('巽', new boolean[] {false , true  , true  } ),
  坎('坎', new boolean[] {false , true  , false } ),
  艮('艮', new boolean[] {false , false , true  } ),
  坤('坤', new boolean[] {false , false , false } );
  
  private char name;
  private boolean[] yinYangs = new boolean[3];
  
  private static final Symbol[] symbolArray = {乾 , 兌 , 離 , 震 , 巽 , 坎 , 艮 , 坤};
  
  private Symbol(char name , boolean[] yinYangs)
  {
    this.name = name;
    this.yinYangs = yinYangs;
  }
  
  /**
   * 取得八個卦
   */
  @NotNull
  public static Symbol[] getSymbols()
  {
    return symbolArray;
  }
  
  @NotNull
  public String toString()
  {
    return String.valueOf(name);
  }
  
  /**
   * 實作 SymbolIF
   * 取得八卦名
   */
  public char getName()
  {
    return name;
  }

  /**
   * 實作 SymbolIF
   * 取得一個卦的第幾爻
   * index 值為 1,2,或3
   */
  @Override
  public boolean getBooleanValue(int index) {
    return yinYangs[index-1];
  }


  /**
   * 「由下而上」 三個陰陽 , 查詢卦象為何
   */
  @NotNull
  private static Symbol getSymbol(YinYangIF[] line)
  {
    for (Symbol aSymbolArray : symbolArray) {
      if ((line[0].getBooleanValue() == aSymbolArray.yinYangs[0]) &&
        (line[1].getBooleanValue() == aSymbolArray.yinYangs[1]) &&
        (line[2].getBooleanValue() == aSymbolArray.yinYangs[2])) {
        return aSymbolArray;
      }
    }
    throw new RuntimeException("Cannot find Symbol for " + line[0] + " , " + line[1] + " , " + line[2]);
  }
  
  /**
   * 「由下而上」 三個陰陽 , 查詢卦象為何
   */
  @NotNull
  public static Symbol getSymbol(boolean[] lines)
  {
    for (Symbol aSymbolArray : symbolArray) {
      if ((lines[0] == aSymbolArray.yinYangs[0]) &&
        (lines[1] == aSymbolArray.yinYangs[1]) &&
        (lines[2] == aSymbolArray.yinYangs[2])) {
        return aSymbolArray;
      }
    }
    throw new RuntimeException("Cannot find Symbol for " + lines[0] + " , " + lines[1] + " , " + lines[2]);
  }

  /**
   * 由 三個陰陽 , 查詢卦象為何 , 比較標準的命名方式
   */
  @NotNull
  public static Symbol valueOf(YinYangIF[] yinYangs)
  {
    return getSymbol(yinYangs);
  }
  
  /**
   * implements FiveElementIF 
   */
  @Nullable
  public FiveElement getFiveElement()
  {
    switch ( this )
    {
      case 乾 : case 兌 : return FiveElement.金 ;
      case 離 :          return FiveElement.火 ;
      case 震 : case 巽 : return FiveElement.木 ;
      case 坎 :          return FiveElement.水 ;
      case 艮 : case 坤 : return FiveElement.土 ;
      default   : return null;
    }
  }
}

File: src/main/java/destiny/iching/Symbol.kt
/**
 * @author smallufo
 * Created on 2002/8/13 at 下午 01:08:17
 */
package destiny.iching

import destiny.core.chinese.FiveElement
import destiny.core.chinese.IFiveElement
import destiny.core.chinese.IYinYang
import java.io.Serializable


/**
 * 八卦基本符號以及其資料
 */
enum class Symbol(private val yinYangs: BooleanArray) : Serializable, ISymbol, IFiveElement {
  乾(booleanArrayOf(true, true, true)),
  兌(booleanArrayOf(true, true, false)),
  離(booleanArrayOf(true, false, true)),
  震(booleanArrayOf(true, false, false)),
  巽(booleanArrayOf(false, true, true)),
  坎(booleanArrayOf(false, true, false)),
  艮(booleanArrayOf(false, false, true)),
  坤(booleanArrayOf(false, false, false));

  override fun toString(): String {
    return name
  }

  /**
   * 取得八卦名
   */
  fun getName(): String {
    return name
  }

  /**
   * 實作 [ISymbol]
   * 取得一個卦的第幾爻
   * index 值為 1,2,或3
   */
  override fun getBooleanValue(index: Int): Boolean {
    return yinYangs[index - 1]
  }


  override val fiveElement: FiveElement
    get() = when (this) {
      乾, 兌 -> FiveElement.金
      離 -> FiveElement.火
      震, 巽 -> FiveElement.木
      坎 -> FiveElement.水
      艮, 坤 -> FiveElement.土
    }


  /** 先天八卦 -> 後天八卦  */
  fun toAcquired(): Symbol {
    return con2AcqMap[this]!!
  }

  /** 後天八卦 -> 先天八卦  */
  fun toCongenital(): Symbol {
    return con2AcqMap.map { (k, v) -> v to k }.toMap()[this]!!
  }

  companion object {

    /** 取得八個卦 */
    private val symbols = arrayOf(乾, 兌, 離, 震, 巽, 坎, 艮, 坤)

    /** 先天八卦 -> 後天八卦 bi-mapping  */
    private val con2AcqMap = mapOf(
      乾 to 離,
      兌 to 巽,
      離 to 震,
      震 to 艮,
      巽 to 坤,
      坎 to 兌,
      艮 to 乾,
      坤 to 坎)


    /**
     * 「由下而上」 三個陰陽 , 查詢卦象為何
     */
    fun getSymbol(line: Array<IYinYang>): Symbol {
      return getSymbol(line[0].booleanValue, line[1].booleanValue, line[2].booleanValue)
    }

    fun of(yinYangs: List<IYinYang>): Symbol {

      return symbols.first {
        it.yinYangs[0] == yinYangs[0].booleanValue &&
          it.yinYangs[1] == yinYangs[1].booleanValue &&
          it.yinYangs[2] == yinYangs[2].booleanValue
      }
    }

    /**
     * 「由下而上」 三個陰陽 , 查詢卦象為何
     */
    fun getSymbol(vararg lines: Boolean): Symbol {
      return symbols.first {
        lines[0] == it.yinYangs[0] &&
          lines[1] == it.yinYangs[1] &&
          lines[2] == it.yinYangs[2]
      }
    }

    /**
     * 由 三個陰陽 , 查詢卦象為何 , 比較標準的命名方式
     */
    fun valueOf(yinYangs: Array<IYinYang>): Symbol {
      return getSymbol(yinYangs)
    }
  }
}
